### PR TITLE
Fix set nofile rlimit error

### PR DESCRIPTION
**Pull request by lifubang**
Fix: #4195 
Close: #4237 

## 1. Fix a get/set race between `runc exec` and `syscall.rlimit.init()`
As @ls-ggg has given the detailed steps to reproduce the issue #4195 , and has given the core reason is that the go runtime will cache the value of rlimit nofile. [1] 
When we are running `runc exec` with nofile limit, runc set it in the parent process, it may cause a race with go runtime init. The race condition is in the time when runc parent process set the container child process's nofile limit just after go runtime init has fetched the nofile limit. [2]
So, we should set nofile limit after `syscall.rlimit.init()` completed.

## 2. Fix an edge case caused by nofile rlimit cache in go stdlib
As @kolyshkin have found an edge case for set nofile rlimit by runc, which is also caused by nofile rlimit cache in go stdlib. Although we have a way to resolve the above get/set race, but if the hard value of nofile rlimit configured in `config.json` is bigger than this value of `runc create/run/exec`, it will also be incorrect restored by `syscall.Exec` in `runc init`. So we need to clear the nofile rlimit cache before we start the container initial process if we need to set nofile rlimit for the container.

[1] https://github.com/golang/go/commit/f5eef58e4381259cbd84b3f2074c79607fb5c821#diff-ec665e9789f8cf5cd1828ad7fa9f0ff4ebc1f5b5dd0fc82a296da5c07da7ece6
[2] https://github.com/golang/go/blob/f5eef58e4381259cbd84b3f2074c79607fb5c821/src/syscall/rlimit.go#L34-L35

### CODE REVIEW

**Comment by lifubang** [comment]
@ls-ggg PTAL, could you please help to test whether this PR fix the issue #4195 or not? Thanks very much.
I has test it for a long time, it seems that the issue has gone with this patch.

This is my test with @kolyshkin 's script in https://github.com/opencontainers/runc/issues/4195#issuecomment-2033428757
```
[2024年 05月 01日 星期三 00:50:14 CST]      82500
[2024年 05月 01日 星期三 00:50:18 CST]      82600
```

File: libcontainer/process_linux.go [hunk 1]
```diff
@@ -340,6 +346,9 @@ func (p *setnsProcess) start() (retErr error) {
 	if err := p.comm.syncSockParent.Shutdown(unix.SHUT_WR); err != nil && ierr == nil {
 		return err
 	}
+	if !seenProcReady && ierr == nil {
+		ierr = errors.New("procReady not received")
+	}
```
Comment (kolyshkin) [inline libcontainer/process_linux.go:351]: I think you don't need this bit.

**Review by kolyshkin — COMMENTED** [review]
Funny, I read your comment and wrote the very same code yesterday, still testing it since it needs about 10 minutes to repro the issue.

File: libcontainer/process_linux.go (hunk 1 above)
Comment (cyphar) [reply libcontainer/process_linux.go:351]: This just mirrors `runc create`. If the `runc init` process dies before sending `procReady` we can detect it here (`parseSync` just waits for `EOF`, it doesn't care about if the process died). AFAICS there are no other liveliness checks after `start()` returns.

File: tests/integration/exec.bats [hunk 2]
```diff
@@ -476,3 +476,15 @@ EOF
...
+	update_config '.process.capabilities.bounding = ["CAP_SYS_RESOURCE"]'
+	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'
+
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
+	[ "$status" -eq 0 ]
+
+	# issue: https://github.com/opencontainers/runc/issues/4195
+	runc exec test_busybox /bin/sh -c "ulimit -n"
+	[[ "${output}" == "65536" ]]
+}
```
Comment (cyphar) [inline tests/integration/exec.bats:490]: As @kolyshkin said in #4266, there isn't much point making a CI test for this -- because this is a race we would need to run this for many iterations to confirm whether the race still exists. Running it once just wastes CI time because even if the bug was re-introduced we would never catch it in a single CI run.

Comment (lifubang) [reply tests/integration/exec.bats]: OK, I'll remove it later, at this time, I used it to test some other issues.

File: tests/integration/exec.bats [hunk 3]
```diff
@@ -476,3 +476,19 @@ EOF
...
+@test "runc exec with RLIMIT_NOFILE" {
+	update_config '.process.capabilities.bounding = ["CAP_SYS_RESOURCE"]'
+	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'
+
+	runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
+	[ "$status" -eq 0 ]
+
+	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
+	runc run test_ulimit
+	[[ "${output}" == "65536" ]]
```
Comment (lifubang) [inline tests/integration/exec.bats:489]: Intrestingly, I don't know how this PR impacts `runc run` with `RLIMIT_NOFILE` in `centos7`:
```
not ok 91 runc exec with RLIMIT_NOFILE
# (in test file tests/integration/exec.bats, line 489)
#   `[[ "${output}" == "65536" ]]' failed
#
# runc run -d --console-socket /tmp/bats-run-PQBrqc/runc.wbkYSh/tty/sock test_busybox (status=0):
#
# runc run test_ulimit (status=0):
# 1024
```

As it will not fail in #4267 .

@kolyshkin @cyphar @AkihiroSuda Do you know why?

**Comment by lifubang** [comment]
It seems that the race situation is that when golang is calling ‘ origRlimitNofile.Load()’ in ‘unix.Exec’, the init function in ‘syscall.rlimit’ will be skipped randomly, but I can’t reproduce in regular code, it seems that this only occurs in runc, maybe related to ‘clone(2)’ in cgo too? I’ll try to write an example tomorrow.

But this PR seems like indeed fixed the Nofile limit set issue in runc, because the error of Setrlimit is ignored in  ‘syscall.Exec’: https://github.com/golang/go/blob/go1.22.2/src/syscall/exec_unix.go#L284-L287

**Comment by lifubang** [comment]
> the init function in ‘syscall.rlimit’ will be skipped randomly

May be there is an error when calling Getrlimit in the init function?
https://github.com/golang/go/blob/go1.22.2/src/syscall/rlimit.go#L30

File: tests/integration/exec.bats (hunk 3 above)
Comment (kolyshkin) [reply tests/integration/exec.bats:489]: @lifubang first, there's no such code (of running a second container with modified `.process.args`) in #4267.

Second, you forgot to check the exit code (here and below).

Third, I think you don't have to add CAP_SYS_RESOURCE as this is the capability of the container itself, and rlimits are set by runc which already have it.

Finally, it looks like you just found a new bug -- `RLIMIT_NOFILE` is not being set in CentOS 7 for `runc run`. Can confirm this

Comment (kolyshkin) [reply tests/integration/exec.bats:489]: @lifubang the issue is, on CentOS 7 the hard limit for NOFILE is 4096. For some reason yet unknown to me, raising the limit is not possible for runc via prlimit(2) syscall, and it returns no error. Maybe it's a bug in CentOS 7 kernel (related to namespaces I guess).

Two ways to workaround it:
1. Use lower limits (soft <= 1024, hard <= 4096) in the test code.
2. Add something like this to the test:

```diff
function setup() {
+	prlimit --nofile=99999:99999 -p $$
	setup_busybox
}
```

Surely, I won't recommend item 2, but it works.

Comment (kolyshkin) [reply tests/integration/exec.bats:491]: Same in Ubuntu 20.04 LTS (kernel 5.15) and Fedora 40 (kernel 6.8) -- if the current hard limit is equal or lower than the one we're trying to set using prlimit syscall, it returns no error but the limit is not changed.

To repro this locally:
```console
$ sudo
# ulimit -H -n 65536
# bats tests/integration/ulimit.bats
root@kir-tp1:/home/kir/git/runc# bats tests/integration/ulimit.bats 
ulimit.bats
 ✗ runc run with RLIMIT_NOFILE
   (in test file tests/integration/ulimit.bats, line 19)
     `[[ "${output}" == "65536" ]]' failed
   runc spec (status=0):
   
   runc run test_ulimit (status=0):
   1024
 ✓ runc exec with RLIMIT_NOFILE
 ✗ runc run+exec two containers with RLIMIT_NOFILE
   (in test file tests/integration/ulimit.bats, line 45)
     `[[ "${output}" == "65536" ]]' failed
   runc spec (status=0):
   
   runc run -d --console-socket /tmp/bats-run-TSa3k7/runc.jz4uuS/tty/sock test_busybox (status=0):
   
   runc run test_ulimit (status=0):
   1024

3 tests, 2 failures
```

Here's the test code (`tests/integration/ulimit.bats`):
```bash
#!/usr/bin/env bats

load helpers

function setup() {
	setup_busybox
}

function teardown() {
	teardown_bundle
}

@test "runc run with RLIMIT_NOFILE" {
	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'

	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
	runc run test_ulimit
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]
}

@test "runc exec with RLIMIT_NOFILE" {
	update_config '	 .process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 2000, "soft": 1000}]'

	runc run -d --console-socket "$CONSOLE_SOCKET" nofile
	[ "$status" -eq 0 ]

	for ((i = 0; i < 100; i++)); do
		runc exec nofile sh -c 'ulimit -n'
		echo "[$i] $output"
		[[ "${output}" == "1000" ]]
	done
}

@test "runc run+exec two containers with RLIMIT_NOFILE" {
	update_config '.process.capabilities.bounding = ["CAP_SYS_RESOURCE"]'
	update_config '.process.rlimits = [{"type": "RLIMIT_NOFILE", "hard": 65536, "soft": 65536}]'

	runc run -d --console-socket "$CONSOLE_SOCKET" test_busybox
	[ "$status" -eq 0 ]

	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
	runc run test_ulimit
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]

	# issue: https://github.com/opencontainers/runc/issues/4195
	runc exec test_busybox /bin/sh -c "ulimit -n"
	[ "$status" -eq 0 ]
	[[ "${output}" == "65536" ]]
}
```
If you set hard ulimit -n to 65537, it works again.

File: libcontainer/process_linux.go (hunk 1 above)
Comment (kolyshkin) [reply libcontainer/process_linux.go:351]: Makes sense.

File: libcontainer/process_linux.go [hunk 4]
```diff
@@ -268,20 +268,26 @@ func (p *setnsProcess) start() (retErr error) {
...
 
+	var seenProcReady bool
 	ierr := parseSync(p.comm.syncSockParent, func(sync *syncT) error {
 		switch sync.Type {
 		case procReady:
-			// This shouldn't happen.
-			panic("unexpected procReady in setns")
+			seenProcReady = true
+			// set rlimits, this has to be done here because we lose permissions
+			// to raise the limits once we enter a user-namespace
```
Comment (kolyshkin) [inline libcontainer/process_linux.go:282]: nit: while at it:
```suggestion
// Set rlimits, this has to be done here because we lose permissions
// to raise the limits once we enter a user-namespace.
```

File: tests/integration/exec.bats [hunk 5]
```diff
@@ -3,6 +3,7 @@
 load helpers
 
 function setup() {
+	[ $EUID -eq 0 ] && prlimit --nofile=1024:1048576 -p $$
```
Comment (kolyshkin) [inline tests/integration/exec.bats:6]: This is ugly and although I've mentioned it (for completeness), I don't recommend doing it, as it is a global setting which affects all the tests down the line.

Easier and cleaner to set lower limits in the tests.

File: tests/integration/exec.bats (hunk 3 above)
Comment (kolyshkin) [reply tests/integration/exec.bats:491]: Here's another addition: we do have a test case for setting RLIMIT_NOFILE during runc exec, but the test case is masking the issue because it sets Cur and Max to the same value (and so Go runtime does not do any tricks with saving/restoring NOFILE).

With this change, the test fails:
```diff
diff --git a/libcontainer/integration/exec_test.go b/libcontainer/integration/exec_test.go
index 9dd056ad..f4b8b839 100644
--- a/libcontainer/integration/exec_test.go
+++ b/libcontainer/integration/exec_test.go
@@ -140,7 +140,7 @@ func testRlimit(t *testing.T, userns bool) {
        // the Setrlimit call happens early enough that we still have permissions to raise the limit.
        ok(t, unix.Setrlimit(unix.RLIMIT_NOFILE, &unix.Rlimit{
                Max: 1024,
-               Cur: 1024,
+               Cur: 512,
        }))
 
        out := runContainerOk(t, config, "/bin/sh", "-c", "ulimit -n")
```

With the above change, the test now fails like this:

```console
[kir@kir-tp1 runc]$ go1.20.14 test -exec sudo  -v libcontainer/integration -run Rlimit
package libcontainer/integration is not in GOROOT (/home/kir/sdk/go1.20.14/src/libcontainer/integration)
[kir@kir-tp1 runc]$ go1.20.14 test -exec sudo  -v ./libcontainer/integration -run Rlimit
=== RUN   TestRlimit
    exec_test.go:148: expected rlimit to be 1025, got 512
--- FAIL: TestRlimit (0.11s)
=== RUN   TestUsernsRlimit
    exec_test.go:148: expected rlimit to be 1025, got 512
--- FAIL: TestUsernsRlimit (0.10s)
=== RUN   TestExecInUsernsRlimit
    execin_test.go:129: expected rlimit to be 1026, got 512
--- FAIL: TestExecInUsernsRlimit (0.14s)
=== RUN   TestExecInRlimit
    execin_test.go:129: expected rlimit to be 1026, got 512
--- FAIL: TestExecInRlimit (0.15s)
FAIL
FAIL	github.com/opencontainers/runc/libcontainer/integration	0.542s
FAIL
```

File: tests/integration/exec.bats (hunk 5 above)
Comment (lifubang) [reply tests/integration/exec.bats]: But in some environments, the soft and hard value of nofile limit are equal. I moved all these type tests to `rlimits.bat`.

File: tests/integration/rlimits.bats [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+@test "runc run with RLIMIT_NOFILE(The same as system's hard value)" {
+	# https://github.com/opencontainers/runc/pull/4265#discussion_r1588599809
+	hard=$(ulimit -n -H)
+	update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"hard\": ${hard}, \"soft\": ${hard}}]"
+	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
+
+	runc run test_ulimit
+	[ "$status" -eq 0 ]
+	[[ "${output}" == "${hard}" ]]
```
Comment (kolyshkin) [inline tests/integration/rlimits.bats:24]: I would add a variable `soft` here, just for the readability.

```bash
	hard=$(ulimit -n -H)
	soft="$hard"
	update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"soft\": ${soft}, \"hard\": ${hard}}]"
	....
	[[ "${output}" == "${soft}" ]]
```

File: libcontainer/standard_init_linux.go [hunk 7]
```diff
@@ -6,17 +6,16 @@ import (
...
 	"github.com/opencontainers/runc/libcontainer/apparmor"
 	"github.com/opencontainers/runc/libcontainer/configs"
 	"github.com/opencontainers/runc/libcontainer/keys"
 	"github.com/opencontainers/runc/libcontainer/seccomp"
 	"github.com/opencontainers/runc/libcontainer/system"
 	"github.com/opencontainers/runc/libcontainer/utils"
+	"github.com/opencontainers/runtime-spec/specs-go"
+	"github.com/opencontainers/selinux/go-selinux"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
```
Comment (kolyshkin) [inline libcontainer/standard_init_linux.go:18]: What is the reason for this change?

Usually in runc code we try to maintain imports grouped into three groups, in order:
- Go standard packages
- vendored packages
- local packages (i.e. github.com/opencontainers/runc)

I think it makes sense to not change this in the existing code.

File: libcontainer/system/linux.go [hunk 8]
```diff
@@ -9,13 +9,29 @@ import (
...
 
+//go:linkname syscallOrigRlimitNofile syscall.origRlimitNofile
+
+// As reported in issue #4195, the new version of go runtime will
+// cache rlimit-nofile. before executing exec, the rlimit-nofile
+// of the process will be restored with the cache. in runc, this will
+// cause the rlimit-nofile set by the parent process for the container
+// to become invalid. this can be solved by clearing the cache. But
+// unfortunately, gostdlib doesn't privede such method, so we need to
+// link to the private var origRlimitNofile in package syscall.
```
Comment (kolyshkin) [inline libcontainer/system/linux.go:28]: nit: why all lowercase?

File: tests/integration/exec.bats (hunk 5 above)
Comment (kolyshkin) [reply tests/integration/exec.bats]: I think I was wrong about the setting being global. Every test case is executed under a separate shell. So it's all right.

**Review by kolyshkin — COMMENTED** [review]
Can you add `runc exec:` prefix to the first commit subject, so it's clear we're patching runc exec? (Alas github does not allow commenting on commit messages)

File: libcontainer/system/linux.go [hunk 9]
```diff
@@ -9,13 +9,29 @@ import (
...
 	"strconv"
+	"sync/atomic"
 	"syscall"
 	"unsafe"
 
 	"github.com/sirupsen/logrus"
 	"golang.org/x/sys/unix"
 )
 
+//go:linkname syscallOrigRlimitNofile syscall.origRlimitNofile
```
Comment (kolyshkin) [inline libcontainer/system/linux.go:20]: While not required, it's better for readability to put go:linkname line just before the variable declaration.

File: libcontainer/setns_init_linux.go [hunk 10]
```diff
@@ -49,6 +49,13 @@ func (l *linuxSetnsInit) Init() error {
 			}
 		}
 	}
+
+	// Set RLIMIT_NOFILE again to clean the cache in go runtime
```
Comment (kolyshkin) [inline libcontainer/setns_init_linux.go:53]: nit: missing period.

File: libcontainer/standard_init_linux.go [hunk 11]
```diff
@@ -77,6 +76,12 @@ func (l *linuxStandardInit) Init() error {
 		}
 	}
 
+	// Set RLIMIT_NOFILE again to clean the cache in go runtime
```
Comment (kolyshkin) [inline libcontainer/standard_init_linux.go:79]: nit: missing period.

**Review by kolyshkin — COMMENTED** [review]
The second patch has `gilibcontainer:` prefix in the subject. Maybe you meant `libcontainer:` or `libct:`

File: tests/integration/rlimits.bats [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n"]'
+
+	runc run test_ulimit
+	[ "$status" -eq 0 ]
+	[[ "${output}" == "${val}" ]]
+}
+
+@test "runc exec with RLIMIT_NOFILE(The same as system's hard value)" {
+	hard=$(ulimit -n -H)
+	update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"hard\": ${hard}, \"soft\": ${hard}}]"
```
Comment (kolyshkin) [inline tests/integration/rlimits.bats:53]: Same here (https://github.com/opencontainers/runc/pull/4265/commits/be7e4169a762b125807701e37fe4843a3dbd9f0b#r1591661042)

File: libcontainer/system/linux.go (hunk 8 above)
Comment (lifubang) [reply libcontainer/system/linux.go]: Thanks.

File: libcontainer/system/linux.go (hunk 9 above)
Comment (lifubang) [reply libcontainer/system/linux.go:20]: Because there should be no space between double slash and `go:linkname`. This looks ugly with other comments before the variable declaration.

Comment (lifubang) [reply libcontainer/system/linux.go:20]: Thanks.

File: libcontainer/setns_init_linux.go [hunk 13]
```diff
@@ -77,6 +77,13 @@ func (l *linuxSetnsInit) Init() error {
 		}
 	}
 
+	// Tell our parent that we're ready to Execv. This must be done before the
```
Comment (kolyshkin) [inline libcontainer/setns_init_linux.go:87]: nit: we do not use `system.Execv` any more (and I'm removing the word in #4268, see https://github.com/opencontainers/runc/pull/4268/commits/bac506463d48faca55baefbeb0e9ec7068088ad3).

Please `s/Execv/exec/` here, too.

File: libcontainer/setns_init_linux.go [hunk 14]
```diff
@@ -49,6 +49,13 @@ func (l *linuxSetnsInit) Init() error {
 			}
 		}
 	}
+
+	// Clean the RLIMIT_NOFILE cache in go runtime.
+	// Issue: https://github.com/opencontainers/runc/issues/4195
+	if containsRlimit(l.config.Rlimits, unix.RLIMIT_NOFILE) {
+		system.ClearRlimitNofileCache()
+	}
+
```
Comment (kolyshkin) [inline libcontainer/setns_init_linux.go:58]: Can you explain why we need the `containsRlimit`? Would be easier to always remove the cache as we don't really need it in runc.

Comment (kolyshkin) [reply libcontainer/setns_init_linux.go:58]: Also, maybe we can move code to a single place, e.g. `system.Exec`.

```diff
diff --git a/libcontainer/system/linux.go b/libcontainer/system/linux.go
index 0b3b4da3..949a5b2a 100644
--- a/libcontainer/system/linux.go
+++ b/libcontainer/system/linux.go
@@ -9,6 +9,7 @@ import (
        "os"
        "os/exec"
        "strconv"
+       "sync/atomic"
        "syscall"
        "unsafe"
 
@@ -44,7 +45,18 @@ func Execv(cmd string, args []string, env []string) error {
        return Exec(name, args, env)
 }
 
+//go:linkname syscallOrigRlimitNofile syscall.origRlimitNofile
+var syscallOrigRlimitNofile atomic.Pointer[syscall.Rlimit]
+
 func Exec(cmd string, args []string, env []string) error {
+       // As reported in issue 4195, the new version of go runtime (since
+       // 1.19) caches rlimit-nofile on startup, and restores it in Exec.
+       // In runc, this causes the rlimit-nofile setting by the parent
+       // process for the container to become invalid. The solution is to
+       // clear this cache (for that, we need a go:linkname hack to access
+       // the private var `origRlimitNofile` in the syscall package).
+       syscallOrigRlimitNofile.Store(nil)
+
        for {
                err := unix.Exec(cmd, args, env)
                if err != unix.EINTR {
```

(i've also modified the comment a bit, for clarity).

File: tests/integration/rlimits.bats [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+	runc run test_ulimit
+	[ "$status" -eq 0 ]
+	[[ "${output}" == "${soft}" ]]
+}
+
+@test "runc run with RLIMIT_NOFILE(Bigger than system's hard value)" {
+	requires root
+	# https://github.com/opencontainers/runc/pull/4265#discussion_r1588599809
+	hard=$(ulimit -n -H)
+	soft=$((hard + 1))
```
Comment (kolyshkin) [inline tests/integration/rlimits.bats:32]: Maybe something like
```bash
lim=$(ulimit -n -H)
hard=$((lim+1))
soft="$hard"
update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"soft\": ${soft}, \"hard\": ${hard}}]"
```
So it's clear what we set.

Also, we can move this code into a separate function to avoid repetition.

E.g.

```bash
# Sets and check rlimit_nofile for runc run. Arguments are:
#  $1: soft limit;
#  $2: hard limit.
function run_check_nofile() {
	soft="$1"
	hard="$2"
	update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"sioft\": ${soft}, \"hard\": ${hard}}]"
	update_config '.process.args = ["/bin/sh", "-c", "ulimit -n; ulimit -H -n"]'
	
	runc run test_rlimit
	[ "$status" -eq 0 ]
	[[ "${lines[0]}" == "${soft}" ]]
	[[ "${lines[1]}" == "${hard}" ]]
}

# Sets and check rlimit_nofile for runc exec. Arguments are:
#  $1: soft limit;
#  $2: hard limit.
function exec_check_nofile() {
	soft="$1"
	hard="$2"
	update_config ".process.rlimits = [{\"type\": \"RLIMIT_NOFILE\", \"sioft\": ${soft}, \"hard\": ${hard}}]"
	
	runc run test_rlimit
	[ "$status" -eq 0 ]
	
	runc exec test_busybox /bin/sh -c "ulimit -n; ulimit -H -n"
	[ "$status" -eq 0 ]
	[[ "${lines[0]}" == "${soft}" ]]
	[[ "${lines[1]}" == "${hard}" ]]
}
```

and call it from all tests:
```bash
val=$(ulimit -n -H)
run_check_nofile "$val" "$val"
```

File: libcontainer/setns_init_linux.go (hunk 14 above)
Comment (lifubang) [reply libcontainer/setns_init_linux.go:58]: > Can you explain why we need the `containsRlimit`? Would be easier to always remove the cache as we don't really need it in runc.

Because if we run a container without a specific nofile rlimit, we should not clear the origRlimitNofile. Consider the system's [curr, max]=[1024, 65536], after we run `runc exec/run/create`, the [curr, max] has been changed to [65536, 65536] by go runtime, and the origRlimitNofile stores [1024, 65536], if we clear this cache, the nofile limit for the container will become wrong.

Comment (lifubang) [reply libcontainer/setns_init_linux.go:58]: > Also, maybe we can move code to a single place, e.g. `system.Exec`.

Consider the above reason, it is difficulty to move code to here.

Comment (kolyshkin) [reply libcontainer/setns_init_linux.go:58]: I think I wrote a comment earlier but it got lost.

We can put this call to `containerInit()`:
```diff
diff --git a/libcontainer/init_linux.go b/libcontainer/init_linux.go
index bd89b253..523259d4 100644
--- a/libcontainer/init_linux.go
+++ b/libcontainer/init_linux.go
@@ -223,6 +223,10 @@ func containerInit(t initType, config *initConfig, pipe *syncSocket, consoleSock
                return err
        }
 
+       if containsRlimit(config.Rlimits, unix.RLIMIT_NOFILE) {
+               system.ClearRlimitNofileCache()
+       }
+
        switch t {
        case initSetns:
                i := &linuxSetnsInit{
```

Comment (lifubang) [reply libcontainer/setns_init_linux.go]: Thanks.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by AkihiroSuda — APPROVED** [review]
